File: 36-valid-sudoku/36-valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {
  const rows = {};
    const cols = {};
    const squares = {};
for(let r = 0; r < 9; r ++){
    for(let c = 0; c < 9; c ++){
        
        const num = board[r][c]
        
       
            if (num === '.') {
                continue;
            }

        
 const grid = `${Math.floor(r / 3)}${Math.floor(c / 3)}`;
        
        if(!rows[r]) {
            rows[r] = new Set()
        }
         if(!cols[c]) {
            cols[c] = new Set()
        }
         if(!squares[grid]) {
            squares[grid] = new Set()
        }
        
        if( rows[r].has(num) || cols[c].has(num) || squares[grid].has(num) ) 
        {
            return false
        }
       
            rows[r].add(num);
          cols[c].add(num);
            squares[grid].add(num);
    }  
}
    
    return true;

}

File: 36-valid-sudoku/36-valid-sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        col = collections.defaultdict(set)
        row = collections.defaultdict(set)
        square = collections.defaultdict(set)
        
        for r in range(9):
            for c in range(9):
                num = board[r][c]
                if num == ".":
                    continue
                grid = (r//3,c//3)
                          
                if (num in row[r] or num in col[c] or num in square[grid]): 
                    return False
                else:
                    row[r].add(num)
                    col[c].add(num)
                    square[grid].add(num)
                        
        return True
                        
                        
                   